lib/srv/git/github.go

```diff
@@ -30,35 +34,141 @@ import (
 
+	"github.com/gravitational/teleport"
 	integrationv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/integration/v1"
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/cryptosuites"
+	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/sshutils"
 )
 
-// knownGithubDotComFingerprints contains a list of known GitHub fingerprints.
-//
-// https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints
-//
-// TODO(greedy52) these fingerprints can change (e.g. GitHub changed its RSA
-// key in 2023 because of an incident). Instead of hard-coding the values, we
-// should try to periodically (e.g. once per day) poll them from the API.
-var knownGithubDotComFingerprints = []string{
-	"SHA256:uNiVztksCsDhcc0u9e8BujQXVUpKZIDTMczCvj3tD2s",
-	"SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM",
-	"SHA256:+DiY3wvvV6TuJJhbpZisF/zLDA0zPMSvHdkr4UvCOqU",
+type githubMetadataClient interface {
+	fetchETag() (string, error)
+	fetchFingerprints() ([]string, string, error)
 }
 
-// VerifyGitHubHostKey is an ssh.HostKeyCallback that verifies the host key
-// belongs to "github.com".
-func VerifyGitHubHostKey(_ string, _ net.Addr, key ssh.PublicKey) error {
+// githubFingerprintManager downloads SSH fingerprints from the GitHub meta API
+// and does a lazy refresh every hour. The fingerprints are used to verify
+// GitHub server when forwarding Git commands to it.
+type githubFingerprintManager struct {
+	mu           sync.RWMutex
+	fingerprints []string
+	lastCheck    time.Time
+	etag         string
+
+	clock  clockwork.Clock
+	client githubMetadataClient
+}
+
+func newGithubFingerprintManager() *githubFingerprintManager {
+	return &githubFingerprintManager{
+		clock:  clockwork.NewRealClock(),
+		client: newGithubMetadataTTPClient(),
+	}
+}
+
+func (g *githubFingerprintManager) checkServerKey(key ssh.PublicKey) error {
 	actualFingerprint := ssh.FingerprintSHA256(key)
-	if slices.Contains(knownGithubDotComFingerprints, actualFingerprint) {
-		return nil
+	for _, fingerprint := range g.getKnownFingerprints() {
+		if sshutils.EqualFingerprints(actualFingerprint, fingerprint) {
+			return nil
+		}
 	}
 	return trace.BadParameter("cannot verify github.com: unknown fingerprint %v algo %v", actualFingerprint, key.Type())
 }
 
+func (g *githubFingerprintManager) getKnownFingerprints() []string {
+	const refreshDuration = time.Hour
+	g.mu.RLock()
+	if g.clock.Now().Sub(g.lastCheck) < refreshDuration {
+		defer g.mu.RUnlock()
+		return g.fingerprints
+	}
+	g.mu.RUnlock()
+
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	if g.clock.Now().Sub(g.lastCheck) < refreshDuration {
+		return g.fingerprints
+	}
+
+	logger := slog.With(teleport.ComponentKey, "git:github")
+
+	// Check if eTag is the same to avoid downloading the whole thing which
+	// contains a lot of irrelevant info.
+	if g.etag != "" {
+		etag, err := g.client.fetchETag()
```

Comment:
This might be a moot point if we switch to the SDK but couldn't we use `If-None-Match` here with a single GET instead of a HEAD and then a GET?